File: src/search/binary_search.py
def search(arr, target):
    left, right = 0, len(arr) - 1
    ans = -1
    while left <= right:
        mid = ((right - left) >> 1) + left
        if arr[mid] < target:
            left = mid + 1
        elif arr[mid] > target:
            right = mid - 1
        else:
            ans = mid
            break
    return ans


arr = [i for i in range(0, 100)]
assert(arr.index(78) == search(arr, 78))

File: src/recursion/BinarySearch.java
package recursion;

public class BinarySearch {
    public static int binarySearch(int[] arr, int target) {
        return process(arr, 0, arr.length - 1, target);
    }

    private static int process(int[] arr, int l, int r, int target) {
        int mid = ((r - l) >> 1) + l;
        // base case
        if (l > r) return -1;
        if (arr[mid] < target) {
            return process(arr, mid + 1, r, target);
        } else if (arr[mid] > target) {
            return process(arr, l, mid, target);
        } else {
            return mid;
        }
    }
}
